miasm2/ir/translators/z3_ir.py

```diff
@@ -0,0 +1,308 @@
+    """
+
+    # Implemented language
+    __LANG__ = "z3"
+    # Operations translation
+    trivial_ops = ["+", "-", "/", "%", "&", "^", "|", "*", "<<"]
+    _cache = None
+    _mem = None
+
+    @classmethod
+    def from_ExprInt(cls, expr):
+        return z3.BitVecVal(expr.arg.arg, expr.size)
+
+    @classmethod
+    def from_ExprId(cls, expr):
+        return z3.BitVec(expr.name, expr.size)
+
+    @classmethod
+    def from_ExprMem(cls, expr):
+        addr = cls.from_expr(expr.arg)
+        return cls._mem.get(addr, expr.size)
+
+    @classmethod
+    def from_ExprSlice(cls, expr):
+        res = cls.from_expr(expr.arg)
+        res = z3.Extract(expr.stop-1, expr.start, res)
+        return res
+
+    @classmethod
+    def from_ExprCompose(cls, expr):
+        res = None
+        args = sorted(expr.args, key=operator.itemgetter(2)) # sort by start off
+        for subexpr, start, stop in args:
+            sube = cls.from_expr(subexpr)
+            e = z3.Extract(stop-start-1, 0, sube)
+            if res:
+                res = z3.Concat(e, res)
+            else:
+                res = e
+        return res
+
+    @classmethod
+    def from_ExprCond(cls, expr):
+        cond = cls.from_expr(expr.cond)
+        src1 = cls.from_expr(expr.src1)
+        src2 = cls.from_expr(expr.src2)
+        return z3.If(cond != 0, src1, src2)
+
+    @classmethod
+    def from_ExprOp(cls, expr):
+        args = map(cls.from_expr, expr.args)
+        res = args[0]
+        for arg in args[1:]:
+            if expr.op in cls.trivial_ops:
+                res = eval("res %s arg" % expr.op)
+            elif expr.op == ">>":
+                res = z3.LShR(res, arg)
+            elif expr.op == "a>>":
+                res = res >> arg
+            elif expr.op == "a<<":
+                res = res << arg
+            else:
+                raise NotImplementedError("Unsupported OP yet: %s" % expr.op)
+        return res
+
+    @classmethod
+    def from_ExprAff(cls, expr):
+        src = cls.from_expr(expr.src)
+        dst = cls.from_expr(expr.dst)
+        return (src == dst)
+
+    @classmethod
+    def from_expr(cls, expr, endianness="<"):
+        # This mess is just to handle cache and Z3Mem instance management
+        # Might be improved in the future
+        del_cache = False
+        del_mem = False
+        if cls._cache is None:
+            cls._cache = {}
+            del_cache = True
+        if cls._mem is None:
+            cls._mem = Z3Mem(endianness)
+            del_mem = True
+
+        try:
+            if expr in cls._cache:
+                return cls._cache[expr]
+            else:
+                return super(TranslatorZ3, cls).from_expr(expr)
+        finally:
+            # Clean cache and Z3Mem if this call is the root call
+            if del_cache:
+                cls._cache = None
+            if del_mem:
+                cls._mem = None
+
+# Register the class
+Translator.register(TranslatorZ3)
+
+if __name__ == '__main__':
```

Comment:
Miasm tests are located in the `test` directory.

This directory tries to follow the `miasm2` package architecture.
So, your tests could be moved to a `test/ir/translators/z3.py`file.

To add them to the test framework, you have to add the unit test command line in `test/test_all.py`.
For instance:
``` Python
## IR
for script in ["ir2C.py",
               "symbexec.py",
               ]:
    testset += RegressionTest([script], base_dir="ir")

testset += RegressionTest(["z3.py"], base_dir="ir/translators")
```

In addition, the z3 import fail should be handled.
To achieve this, you can define a *tag* for your z3 test, for instance `z3`, and assign it to your test (`tags=[TAGS["z3"]]`).

That way, as LLVM, you can disable the test case by using:
``` Python
try:
    import z3
except ImportError:
    # Print warning / sleep ...        
    if TAGS["z3"] not in exclude_tags:
            exclude_tags.append(TAGS["z3"])
```